### PR TITLE
Fix #7585: Portfolio filters and display settings

### LINKED ISSUE #7585: Portfolio Filters and Display Settings
### Description: 

Add a modal that contains filters and display settings for Portfolio (ideally re-usable with NFT / Activity tabs, etc.).

Filters and Display Settings | Filters and Display Settings
--|--

Group By (TBD, needs Portfolio v2 designs):
- Accounts
- Network
- None

Sort Assets:
- Price/fiat high -> low (default, current behaviour)
- Price/fiat low -> high
- A-Z
- Z-A

Hide assets with <$1 value toggle.

Network Filter(s):
- Checkmark/box beside each network, grouped by Primary/Secondary/Test networks. 
- Primary networks are shown, secondary & test networks hidden behind expandable arrow button.
- All networks selected by default, except test networks (current behaviour).

Accounts Filter:
- All accounts are selected by default

cc @kleantzogu @jamesmudgett @HEagle18

### PR DESCRIPTION
**Pull request by StephenHeaps**
## Summary of Changes
- Added `Filters and Display Settings` modal with multiple filter options & persistence to Portfolio:
    - Change sort, hide small balances, filter by accounts and by networks.
- Refactored `AccountListView` to support multiple selection, similar style to networks. 
    - There is a shared root view, a filter view and a selection view. 
    - `NavigationView` was pulled out so we can push it as a detail view

This pull request fixes #7585

## Submitter Checklist:

- [x] *Unit Tests* are updated to cover new or changed functionality
- [x] User-facing strings use `NSLocalizableString()`
- [x] New or updated UI has been tested across:
  - [x] Light & dark mode
  - [x] Different size classes (iPhone, landscape, iPad)
  - [x] Different dynamic type sizes

## Test Plan:
NOTE: 60% presentation size is a nice to have only available on iOS 16 and later.

### Test `Sort Assets`:
1. Go to Portfolio and tap the filters button.
2. Change the `Sort Assets` option and tap `Save Changes`.
3. The portfolio should reflect your selected option.
4. Test this for all options.

https://github.com/brave/brave-ios/assets/5314553/b8b5536a-7f0a-4110-849e-e07144820494

### Test `Hide Small Balances`:
1. Go to Portfolio and tap the filters button.
2. Toggle on the `Hide Small Balances` toggle and tap `Save Changes`.
3. Verify the portfolio hide assets with a balance less than $1.

https://github.com/brave/brave-ios/assets/5314553/97887ab7-4357-4de4-a967-1d7a7397b3b8

### Test `Select Accounts`
1. Go to Portfolio and tap the filters button.
2. Tap `Select Accounts`.
3. De-select some accounts and tap `Save Changes`.
4. Verify the balances are updated to reflect the new accounts.
    - Tip: If you de-select all Solana accounts, SOL should show as 0.
    - Tip: If you have multiple accounts for one coin type, you can tap to Asset Detail for an asset to see the balance for ALL accounts which can be used to help verify only the selected accounts balance(s) are shown in Portfolio.

https://github.com/brave/brave-ios/assets/5314553/de802b3a-1fe6-4a9b-ac51-e44bf0059733

### Testing `Select Networks`:
1. Go to Portfolio and tap the filters button.
2. Tap `Select Networks`.
3. De-select some networks and tap `Save Changes`.
4. Verify the assets from the de-selected networks are removed.

https://github.com/brave/brave-ios/assets/5314553/ba3d38ee-6a20-49f5-871e-ecc75195ab86

### Test `Reset` & persistence
1. Go to Portfolio and tap the filters button.
2. Change some filters and tap `Save Changes`.
3. Close wallet and/or Brave app
4. Re-open the app to Portfolio,  and tap the filters button.
5. Verify the filters were preserved
6. Tap `Reset`
7. Verify the filters were reset. Defaults:
    - `Sort Assets`: High to Low
    - `Hide Small Balances`: Off/disabled
    - `Select Accounts`: all accounts selected
    - `Select Networks`: all networks except test networks (when enabled) are selected

https://github.com/brave/brave-ios/assets/5314553/4e31ec0f-f822-4129-8192-d7bb04aca086

## Screenshots:

Light Mode | Dark Mode
--|--
Filters   Display Settings (light) | Filters   Display Settings (dark)
Account Filter (light) | Account Filter (dark)
Network Filter (light) | Network Filter (dark)

## Reviewer Checklist:

- [ ] Issues include necessary QA labels:
  - `QA/(Yes|No)`
  - `bug` / `enhancement`
- [ ] Necessary [security reviews](https://github.com/brave/security/issues/new/choose) have taken place.
- [ ] Adequate unit test coverage exists to prevent regressions.
- [ ] Adequate test plan exists for QA to validate (if applicable).
- [ ] Issue and pull request is assigned to a milestone (should happen at merge time).

### CODE REVIEW

File: Sources/BraveWallet/Crypto/Accounts/AccountFilterView.swift [hunk 1]
```diff
@@ -0,0 +1,41 @@
+// Copyright 2022 The Brave Authors. All rights reserved.
```
Comment (nuo-xu) [inline Sources/BraveWallet/Crypto/Accounts/AccountFilterView.swift]: 2023 hehehehe

File: Sources/BraveWallet/Crypto/Accounts/AccountSelectionRootView.swift [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+              if allSelected { // deselect all
+                allAccounts.forEach(selectAccount)
+              } else { // select all
+                let unselectedAccounts = allAccounts
+                  .filter { account in
+                    !selectedAccounts.contains(
+                      where: { $0.id == account.id && $0.coin == account.coin }
+                    )
+                  }
+                unselectedAccounts.forEach(selectAccount)
```
Comment (nuo-xu) [inline Sources/BraveWallet/Crypto/Accounts/AccountSelectionRootView.swift]: can combine these two loops into one?
ForEach(allAccounts) { account in 
  if !selectedAccounts.contains(where: { $0.id == account.id && $0.coin == account.coin }) {
    selectAccount(account)
  }
}

File: Sources/BraveWallet/Crypto/Accounts/AccountSelectionRootView.swift [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+    .listBackgroundColor(Color(uiColor: WalletV2Design.containerBackground))
+    .navigationTitle(navigationTitle)
+    .navigationBarTitleDisplayMode(.inline)
+  }
+}
+
+private struct AccountListRowView: View {
+  
+  var account: BraveWallet.AccountInfo
+  var selectedAccounts: [BraveWallet.AccountInfo]
```
Comment (nuo-xu) [inline Sources/BraveWallet/Crypto/Accounts/AccountSelectionRootView.swift]: i dont think `AccountListRowView` needs to know what `selectedAccounts` we can compute `isSelected` boolean outside of this struct and pass it via init

File: Sources/BraveWallet/Crypto/NetworkSelectionRootView.swift [hunk 4]
```diff
@@ -12,66 +12,204 @@ struct NetworkSelectionRootView: View {
...
+                let unselectedNetworks = allNetworks.testNetworks
+                  .filter { testNetwork in
+                    !selectedNetworks.contains(
+                      where: { $0.chainId == testNetwork.chainId && $0.coin == testNetwork.coin }
+                    )
+                  }
+                unselectedNetworks.forEach(selectNetwork)
+              }
+            }
+          )
```
Comment (nuo-xu) [inline Sources/BraveWallet/Crypto/NetworkSelectionRootView.swift]: feel like this is being shared multiple times. can be refactored to avoid code duplication.

File: Sources/BraveWallet/Crypto/Stores/PortfolioStore.swift [hunk 5]
```diff
@@ -381,3 +428,33 @@ extension PortfolioStore: BraveWalletBraveWalletServiceObserver {
...
+  }
+}
+
+extension Array {
+  /// `filter` helper that skips iterating through the entire array when not applying any filtering.
+  @inlinable public func optionallyFilter(shouldFilter: Bool, isIncluded: (Element) throws -> Bool) rethrows -> [Element] {
+    guard shouldFilter else { return self }
+    return try filter(isIncluded)
+  }
+}
```
Comment (nuo-xu) [inline Sources/BraveWallet/Crypto/Stores/PortfolioStore.swift:474]: is this PR only implement the filter in Portfolio and the NFT one will be in another PR?

Comment (StephenHeaps) [reply Sources/BraveWallet/Crypto/Stores/PortfolioStore.swift:474]: Yes NFT tab will follow after but re-use the same `FilterDisplaySettingsView`. https://github.com/brave/brave-ios/issues/7699

File: Sources/BraveWallet/Crypto/Accounts/AccountSelectionRootView.swift (hunk 3 above)
Comment (StephenHeaps) [reply Sources/BraveWallet/Crypto/Accounts/AccountSelectionRootView.swift]: Thanks, I've also updated `NetworkListRowView` to match in [89405ba](https://github.com/brave/brave-ios/pull/7665/commits/89405ba3ef0c8cb5b55ce9ec8f8f7c07643c9833) 👍.

File: Sources/BraveWallet/Crypto/NetworkSelectionRootView.swift (hunk 4 above)
Comment (StephenHeaps) [reply Sources/BraveWallet/Crypto/NetworkSelectionRootView.swift]: Good idea, I've made a `SelectAllHeaderView` for re-use with this view (primary networks, secondary networks, test networks headers) and with `AccountSelectionRootView` in [89405ba](https://github.com/brave/brave-ios/pull/7665/commits/89405ba3ef0c8cb5b55ce9ec8f8f7c07643c9833)👍.

File: Sources/BraveWallet/Crypto/Stores/PortfolioStore.swift (hunk 5 above)
Comment (amazonXXX) [reply Sources/BraveWallet/Crypto/Stores/PortfolioStore.swift:467]: d31189d50fced292c429e460cf15045fcca584c5

**Review by nuo-xu — APPROVED** [review]
